PeerioTechnologies/peerio-icebear
# feat: parallelize chunk download

For discussion/initial review.

Needs `config.download.parallelism` in clients set to something more than 1 for proper testing.

<s>(Based off #112, will rebase on top of dev after #112 is merged, so ignore `tests` diff here, discuss it there.)</s>

#### Relevant info and issue/PR links 
 
https://app.clubhouse.io/peerio/story/707/parallelize-chunk-download

#### Testing instructions  

(After review)

* Set various `config.download.parallelism` (1, 2, 4) in the client (in Desktop: add `cfg.download.parallelism = 2`; in `config.js`). We'll probably use 2 initially.
* Download lots of different files, from tiny to very large. Make sure they match exactly with the original — we want to make sure chunks are downloaded correctly and put in the right order (e.g. before uploading test files calculate `shasum uploaded-file` and then after downloading do `shasum downloaded-file` and compare hashes — they must be the same).

----
### Repository owner


[PR guideline](https://github.com/PeerioTechnologies/peerio-icebear/blob/dev/docs/CONTRIBUTING.md)
